sdk/dotnet/Pulumi/X/Automation/IWorkspace.cs

```diff
@@ -0,0 +1,198 @@
+        /// </summary>
+        PulumiFn? Program { get; set; }
+
+        /// <summary>
+        /// Environment values scoped to the current workspace. These will be supplied to every Pulumi command.
+        /// </summary>
+        IDictionary<string, string>? EnvironmentVariables { get; set;}
+
+        /// <summary>
+        /// Returns project settings for the current project if any.
+        /// </summary>
+        Task<ProjectSettings?> GetProjectSettingsAsync();
+
+        /// <summary>
+        /// Overwrites the settings for the current project.
+        /// <para/>
+        /// There can only be a single project per workspace. Fails if new project name does not match old.
+        /// </summary>
+        /// <param name="settings">The settings object to save.</param>
+        Task SaveProjectSettingsAsync(ProjectSettings settings);
+
+        /// <summary>
+        /// Returns stack settings for the stack matching the specified stack name if any.
+        /// </summary>
+        /// <param name="stackName">The name of the stack.</param>
+        Task<StackSettings?> GetStackSettingsAsync(string stackName);
+
+        /// <summary>
+        /// Overwrite the settings for the stack matching the specified stack name.
+        /// </summary>
+        /// <param name="stackName">The name of the stack to operate on.</param>
+        /// <param name="settings">The settings object to save.</param>
+        Task SaveStackSettingsAsync(string stackName, StackSettings settings);
+
+        /// <summary>
+        /// Hook to provide additional args to every CLI command before they are executed.
+        /// <para/>
+        /// Provided with a stack name, returns an array of args to append to an invoked command <c>["--config=...", ]</c>.
+        /// <para/>
+        /// <see cref="LocalWorkspace"/> does not utilize this extensibility point.
+        /// </summary>
+        /// <param name="stackName">The name of the stack.</param>
+        Task<ImmutableList<string>> SerializeArgsForOpAsync(string stackName);
+
+        /// <summary>
+        /// Hook executed after every command. Called with the stack name.
+        /// <para/>
+        /// An extensibility point to perform workspace cleanup (CLI operations may create/modify a Pulumi.stack.yaml).
+        /// <para/>
+        /// <see cref="LocalWorkspace"/> does not utilize this extensibility point.
+        /// </summary>
+        /// <param name="stackName">The name of the stack.</param>
+        Task PostCommandCallbackAsync(string stackName);
+
+        /// <summary>
+        /// Returns the value associated with the specified stack name and key, scoped
+        /// to the Workspace.
+        /// </summary>
+        /// <param name="stackName">The name of the stack to read config from.</param>
+        /// <param name="key">The key to use for the config lookup.</param>
+        Task<ConfigValue> GetConfigAsync(string stackName, string key);
```

Comment:
I have some thoughts about the naming of the `Config` related methods on `IWorkspace`. Currently it is:

- `GetConfigAsync` for a single `ConfigValue`.
- `GetAllConfigAsync` for the config map.

The other methods (set, remove) are named similarly. I suggest doing this:

- `GetConfigValueAsync` for a single `ConfigValue`.
- `GetConfigAsync` for the config map.

Just because it will keep the config-related methods near each other for intellisense. Thoughts?

Reply:
I like this suggestion. If we do make a naming change here, we should consider updating the other languages as well to be consistent.

Another option:

- `GetConfigValueAsync` (singular) for a single `ConfigValue`.
- `GetConfigValuesAsync` (plural) for the config map.

Which is similar to the built-in [`Environment.GetEnvironmentVariable`](https://docs.microsoft.com/en-us/dotnet/api/system.environment.getenvironmentvariable?view=net-5.0) and [`Environment.GetEnvironmentVariables`](https://docs.microsoft.com/en-us/dotnet/api/system.environment.getenvironmentvariables?view=net-5.0).

Another option: for languages that support overloads (C#, TypeScript, Python), we could have two overloads with the same name (e.g. `GetConfigAsync`). The overload that accepts just the `stackName` returns all config, the overload that accepts both the `stackName` and `key` returns just the value:

```c#
Task<ImmutableDictionary<string, ConfigValue>> GetConfigAsync(string stackName);
Task<ConfigValue> GetConfigAsync(string stackName, string key);
```

Go doesn't support overloads, so if we move to this approach, we'd need to either keep Go's naming as-is, or rename based on the other naming suggestions above. Or another possible option for Go is `GetConfig` (single value) and `GetConfigMap` (all config).

@EvanBoyle, thoughts?

Reply:
Another idea I've been kicking around, but not sure how I feel about it just yet... grouping functionality together more. Something like:

```c#
public abstract class Workspace
{
    // ...

    public abstract WorkspaceConfig Config { get; }
    public abstract WorkspaceStacks Stacks { get; }
    public abstract WorkspacePlugins Plugins { get; }

    // ...
}

public abstract class WorkspaceConfig
{
    public abstract Task<ConfigValue> GetAsync(string stackName, string key);
    public abstract Task<ImmutableDictionary<string, ConfigValue>> GetAllAsync(string stackName);

    public abstract Task SetAsync(string stackName, string key, ConfigValue value);
    public abstract Task SetAllAsync(string stackName, IDictionary<string, ConfigValue> configMap);

    public abstract Task RemoveAsync(string stackName, string key);
    public abstract Task RemoveAllAsync(string stackName, IEnumerable<string> keys);

    public abstract Task<ImmutableDictionary<string, ConfigValue>> RefreshAsync(string stackName);
}

public abstract class WorkspaceStacks
{
    public abstract Task<StackInfo?> GetSelectedAsync();
    public abstract Task CreateAsync(string stackName);
    public abstract Task SelectAsync(string stackName);
    public abstract Task RemoveAsync(string stackName);
    public abstract Task<ImmutableList<StackInfo>> ListAsync();
}

public abstract class WorkspacePlugins
{
    public abstract Task InstallAsync(string name, string version, string? kind = null);
    public abstract Task RemoveAsync(string? name = null, string? versionRange = null, string? kind = null);
    public abstract Task<ImmutableList<PluginInfo>> ListAsync();
}
```

With usage like:

```c#
Workspace ws = ...;
var config = await ws.Config.GetAllAsync();
await ws.Stacks.SelectAsync("foo");
await ws.Plugins.InstallAsync("aws", "2.0.0");
```

Reply:
@justinvp Interesting idea. One thing I have a hard time with is the relationship between `Workspace`, `Stack`, and `Config`. The example is missing a param for config:

```cs
var config = await ws.Config.GetAllAsync("bar");
await ws.Stacks.SelectAsync("foo");
await ws.Plugins.InstallAsync("aws", "2.0.0");
```

Both config and stack are keyed by stack name, which makes the grouping feel a bit odd. The config belongs to a stack and the relationship with workspace is only really for persistence. Not sure if this subgrouping over-emphasizes the relationship between config and workspace.

I am a fan of using better names for single value vs group operations on config.

Reply:
So if we went with that then it would probably make more sense for `Config` manipulation to be on the `Stacks` object?
```csharp
var config = await ws.Stacks.GetConfigAsync("stack");
var value = await ws.Stacks.GetConfigValueAsync("stack", "key");
```

so at that point is there still value in keeping `Stacks` & `Plugins` separated?

Reply:
> Both config and stack are keyed by stack name, which makes the grouping feel a bit odd. The config belongs to a stack and the relationship with workspace is only really for persistence. Not sure if this subgrouping over-emphasizes the relationship between config and workspace.

Agreed. Yeah, my initial factoring isn't quite right: if we were to go down this path, the config-related things would probably be better on `Stacks`, but even then it doesn't really make it clear that this is config for the current stack.

In any case, apologies for taking us on this tangent. Let's ignore my grouping idea for now and go back to discussing the previous naming suggestions on the top-level workspace:

### Current naming:

- `GetConfigAsync` for a single `ConfigValue`.
- `GetAllConfigAsync` for the config map.

### Option 1 (@orionstudt's original suggestion):

- `GetConfigValueAsync` for a single `ConfigValue`.
- `GetConfigAsync` for the config map.

### Option 2:

- `GetConfigValueAsync` (singular) for a single `ConfigValue`.
- `GetConfigValuesAsync` (plural) for the config map.

Which is similar to the built-in [`Environment.GetEnvironmentVariable`](https://docs.microsoft.com/en-us/dotnet/api/system.environment.getenvironmentvariable?view=net-5.0) and [`Environment.GetEnvironmentVariables`](https://docs.microsoft.com/en-us/dotnet/api/system.environment.getenvironmentvariables?view=net-5.0).

### Option 3:

For languages that support overloads (C#, TypeScript, Python), we could have two overloads with the same name (e.g. `GetConfigAsync`). The overload that accepts just the `stackName` returns all config, the overload that accepts both the `stackName` and `key` returns just the value:

```c#
Task<ImmutableDictionary<string, ConfigValue>> GetConfigAsync(string stackName);
Task<ConfigValue> GetConfigAsync(string stackName, string key);
```

Go doesn't support overloads, so if we move to this approach, we'd need to either keep Go's naming as-is, or rename based on the other naming suggestions above. Or another possible option for Go is `GetConfig` (single value) and `GetConfigMap` (all config).

---

Personally, I'm in favor of Option 1 (@orionstudt's original suggestion). If we do this, we should be prepared to change the naming for TS and Go for consistency.

Reply:
I went with Option 1 for now, can decide more later.

Reply:
Also IDK if you want any naming changes made to the other automation APIs to be done in this PR or a different PR?

Reply:
I think it would be best to keep this change set targeted to C#. Before merging we can open some tracking issues for updates to be made to other SDKs. These will involve updating the samples repo as well, so it is a little less contained.